File: tools/m1_occupancy.py
#!/usr/bin/env python3
"""REBUILD Phase 0 / M1 decoder — arcade 68K frame occupancy.

Reads error.log lines `M1S <beamy hex> <frame hex>` written by
tools/m1_occupancy.lua (a breakpoint on every `stop #$2300` site).

Model: the game halts at a STOP until the vint (line 224) wakes it.
Idle lines for a stop at line V: (224-V) if V<224 else (262-V)+224.
Multiple stops per frame = re-idling after a mid-frame IRQ wake; sum
spans conservatively by taking the FIRST stop only (later stops sit
inside the same pre-vint span). Frames with NO stop = 100% busy.

Occupancy = 1 - idle/262. The M1 question: what does the busy-frame
population look like, especially in gameplay (frames > 1100)?
The MD 68K runs the game at 0.77x arcade clock minus handler share:
a frame needs occupancy <= ~0.72 to fit 60Hz on the port.
"""
import sys
import collections

LINES = 262
VINT = 224

path = sys.argv[1] if len(sys.argv) > 1 else "error.log"
first_stop = {}
hits_per_frame = collections.Counter()
for line in open(path, errors="ignore"):
    if not line.startswith("M1S "):
        continue
    try:
        _, v_hex, f_hex = line.split()
        v, f = int(v_hex, 16), int(f_hex, 16)
    except ValueError:
        continue
    hits_per_frame[f] += 1
    if f not in first_stop or v < first_stop[f]:
        first_stop[f] = v

if not first_stop:
    sys.exit("no M1S hits in the log")

frames = sorted(first_stop)
f_lo, f_hi = frames[0], frames[-1]


def occupancy(f):
    # wait-entry model (hook 0x397E): work runs from the vint (line
    # 224) to the wait entry at line V; multi-frame waits re-enter
    # the hook just after each wake, reading as handler-only frames.
    if f not in first_stop:
        return 1.0
    v = first_stop[f]
    return ((v - VINT) % LINES) / LINES


def report(name, lo, hi):
    occ = [occupancy(f) for f in range(lo, hi)]
    if not occ:
        return
    occ_s = sorted(occ)
    busy100 = sum(1 for f in range(lo, hi) if f not in first_stop)
    over = sum(1 for o in occ if o > 0.72)
    n = len(occ)
    print(f"{name}: frames={n} mean occ={sum(occ)/n*100:.1f}% "
          f"p50={occ_s[n//2]*100:.1f}% p95={occ_s[int(n*.95)]*100:.1f}% "
          f"p99={occ_s[int(n*.99)]*100:.1f}% max={occ_s[-1]*100:.1f}%")
    print(f"  100%-busy frames (no STOP): {busy100} "
          f"({100.0*busy100/n:.2f}%)   frames over the 0.72 port "
          f"budget: {over} ({100.0*over/n:.2f}%)")


print(f"M1 ARCADE OCCUPANCY — stop-hits logged over frames "
      f"{f_lo}..{f_hi}; multi-stop frames: "
      f"{sum(1 for c in hits_per_frame.values() if c > 1)}")
report("ALL      ", f_lo, f_hi)
report("attract  ", f_lo, min(1100, f_hi))
report("gameplay ", 1100, f_hi)

File: tools/m1_occupancy.lua
-- REBUILD Phase 0 / M1 — ARCADE 68K FRAME OCCUPANCY.
-- Altered Beast's main loop idles via `stop #$2300` (12 sites): the
-- CPU halts until the vint wakes it. Breakpoint every STOP site and
-- log (beam line, frame) per entry; idle-per-frame is then the span
-- from the STOP to the next vint (line 224). Frames with NO stop hit
-- are 100%-busy frames — the population that decides whether the
-- original binary can pace 60Hz on the MD's 0.77x 68K.
--   mame altbeast -debug -debugger none -log <headless flags> \
--       -autoboot_script tools/m1_occupancy.lua
-- Decode: tools/m1_occupancy.py (reads error.log).
-- THE canonical idle hook: 0x397E is the vint-wait ENTRY
-- (`clr.b $FFF01C; tst/beq poll`) — executes ONCE per wait period.
-- The 12 `stop #$2300` sites turned out to be other-mode idles (PC
-- sampling showed the attract/game idling at 0x3982); keep them as
-- secondary hooks anyway.
local stops = {0x397e,
               0x1a9e6,0x1b0ca,0x1b178,0x1b1d2,0x1b35a,0x1b4ac,
               0x1b4b0,0x1b50c,0x1b5b2,0x1b5d0,0x1b93a,0x1b9e6}
local frames = 0
local fields = {}
local inited = false
local function init_fields()
    for tag, port in pairs(manager.machine.ioport.ports) do
        for fname, field in pairs(port.fields) do fields[fname] = field end
    end
    inited = true
end
local function set_input(name, val)
    local f = fields[name]
    if f then f:set_value(val) end
end
emu.register_frame_done(function()
    if not inited then init_fields() end
    frames = frames + 1
    if frames == 600 or frames == 800 then set_input('Coin 1', 1) end
    if frames == 640 or frames == 840 then set_input('Coin 1', 0) end
    if frames == 1000 then set_input('1 Player Start', 1) end
    if frames == 1040 then set_input('1 Player Start', 0) end
    if frames > 1100 then
        local p = frames % 240
        set_input('P1 Right', (p < 120) and 1 or 0)
        set_input('P1 Button 1', (p % 40 < 8) and 1 or 0)
    end
    if frames == 120 then
        for _, a in ipairs(stops) do
            manager.machine.debugger:command(string.format(
                'bpset %x,1,{logerror "M1S %%04X %%08X\\n",beamy,frame;g}', a))
        end
    end
    if frames == 5400 then
        manager.machine.debugger:command('bpclear')
        manager.machine:exit()
    end
end)
